docs/decisions/0015-serving-static-assets.rst

```diff
@@ -0,0 +1,125 @@
+
+**Serving assets should not *require* ASGI deployment.**
+  Deploying the LMS and Studio using ASGI would likely substantially improve the scalability of a Django-based streaming solution, but migrating and testing this new deployment type for the entire stack is a large task and is considered out of scope for this project.
+
+Decision
+--------
+
+URLs
+~~~~
+
+The format will be: ``https://{asset_server}/assets/apps/{app}/{learning_package_key}/{component_key}/{version}/{filepath}``
+
+The asset server will have a completely different domain from the LMS and Studio, and will not be a subdomain.
+
+A more concrete example: ``https://studio.assets.sandbox.openedx.io/apps/content_libraries/lib:Axim:200/xblock.v1:problem@826eb471-0db2-4943-b343-afa65a6fdeb5/v2/static/images/fig1.png``
+
+The ``version`` can be:
+
+* ``draft`` indicating the latest draft version (viewed by authors in Studio).
+* ``published`` indicating the latest published version (viewed by students in the LMS)
+* ``v{num}`` meaning a specific version–e.g. ``v20`` for version 20.
+
+Site Separation
+~~~~~~~~~~~~~~~
+
+The asset-serving site will run in the same process as the LMS or Studio process the links, so ``studio.assets.sandbox.openedx.io`` would point to the same place that ``studio.sandbox.openedx.org``. Being in the same process will make it easier to implement permissions checks and URL generation code.
+
+The switching between the two would happen with a new Middleware class that is implemented in openedx-learning. We would use this Middleware to make sure that assets are never served through the Studio and LMS URLs, and that no other Studio or LMS views are served through the asset server URL. This will require new configuration settings.
+
+View Implementation
+~~~~~~~~~~~~~~~~~~~
+
+The Learning Core will implement a Django REST Framework APIView that will provide the core functionality for serving assets. This view will be subclassed and customized by apps in LMS and Studio that want to make use of this functionality.
+
+The Learning Core will also provide auth related endpoints.
+
+Permissions
+~~~~~~~~~~~
+
+The Learning Core will not implement any permissions checking. It will be the responsibility of an app in Studio or the LMS add permissions classes to their subclass of the APIView.
+
+For instance, Studio would likely implement a simple permissions checking model that only examines the learning context and restricts access to course staff. LMS might eventually use a much more sophisticated model that looks at the individual Component that an asset belongs to.
+
+Cookie Authentication
+~~~~~~~~~~~~~~~~~~~~~
+
+Authentication will use a domain cookie on the root assets domain (e.g. ``*.assets.sandbox.openedx.io``). The cookie will have the ``Secure`` and ``HttpOnly`` attributes.
+
+OPEN QUESTIONS:
+
+* What's the best way to do handoff between LMS/Studio and get that cookie information over to the asset domains?
```

Comment:
Maybe:

1. Browser requests a file without any auth headers
2. Edge function requests the metadata from the LMS API (if not already cached) to determine if auth is needed.
3. If no auth is needed, serve the file directly.
4. If auth is needed, return a 302 redirect to a special LMS endpoint. That endpoint redirects back to the CDN but includes an auth header in the query string.
5. Edge function serves the file along with a `Set-Cookie` header, moving the auth token from query string to cookie.